src/streams.h

```diff
@@ -296,6 +296,16 @@ class CDataStream
         data.insert(data.end(), begin(), end());
         clear();
     }
+
+    /**
+     * XOR the contents of this stream with a certain key.
+     *
+     * @param[in] key    The key used to XOR the data in this stream.
+     */
+    void Xor(const std::string& key) {
```

Comment:
The problem is that you use eg. std::string("\000\000\000").
If you do that, C++ will use strlen() to determine the length, which looks for 0 characters.
Pass in a length explicitly.

So do either:

    std::string(3, '\000');  // Create a string with three 0 bytes

Or

    static const char data[] = {0,0,0};
    std::string(data, data+sizeof(data));

...

Don't use the NUL-terminated string constructor for raw data that can contain 0.